generals/remote/generalsio_client.py

```diff
@@ -0,0 +1,109 @@
+from socketio import SimpleClient  # type: ignore
+
+from generals.agents.agent import Agent
+
+
+class GeneralsBotError(Exception):
+    """Base generals-bot exception
+    TODO: find a place for exceptions
+    """
+
+    pass
+
+
+class GeneralsIOClientError(GeneralsBotError):
+    """Base GeneralsIOClient exception"""
+
+    pass
+
+
+class RegisterAgentError(GeneralsIOClientError):
+    """Registering bot error"""
+
+    def __init__(self, msg: str) -> None:
+        super().__init__()
+        self.msg = msg
+
+    def __str__(self) -> str:
+        return f"Failed to register the agent. Error: {self.msg}"
+
+
+class GeneralsIOClient(SimpleClient):
+    """
+    Wrapper around socket.io client to enable Agent to join
+    GeneralsIO lobby.
+    """
+
+    def __init__(self, agent: Agent, user_id: str):
+        super().__init__()
+        self.connect("https://botws.generals.io")
+        self.user_id = user_id
+        self._queue_id = ""
+
+    @property
+    def queue_id(self):
+        if not self._queue_id:
+            raise GeneralsIOClientError("Queue ID is not set.\nIs agent in the game lobby?")
+
+        return self._queue_id
+
+    def _emit_receive(self, *args):
+        self.emit(*args)
+        return self.receive()
+
+    def register_agent(self, username: str) -> None:
+        """
+        Register Agent to GeneralsIO platform.
+        You can configure one Agent per `user_id`. `user_id` should be handled as secret.
+        :param user_id: secret ID of Agent
+        :param username: agent username, must be prefixed with `[Bot]`
+        """
+        event, response = self._emit_receive("set_username", (self.user_id, username))
+        if response:
+            # in case of success the response is empty
+            raise RegisterAgentError(response)
+
+    def join_private_game(self, queue_id: str) -> None:
+        """
+        Join (or create) private game lobby.
+        :param queue_id: Either URL or lobby ID number
+        """
+        self._emit_receive("join_private", (queue_id, self.user_id))
+        self._queue_id = queue_id
+
+    def wait_for_game(self, force_start: bool = True) -> None:
+        """
+        Wait for the game start.
+        :param force_start: If set to True, the Agent will set `Force Start` flag
+        """
+        if force_start:
+            self.emit("set_force_start", (self.queue_id, True))
+
+        while True:
```

Comment:
question/suggestion: given this implementation, if agent is in queue, the loop makes CPU to run on 100% right?
It would be nice if there was something like `time.sleep(0.1)`, wdyt?

Reply:
Not really. The `self.receive()` method is active waiting for a message. As long as there are no new messages, the method blocks the whole thread (basically sleeps, I am not sure how it is implemented, but I suspect it to be similar to sleep method --> block thread execution until interrupt, where interrupt in this case will be a new message). Because of that, the while loop is stuck as well.

Reply:
so is the loop required then?

Reply:
Yes. Without it the method would end after receiving a single message. For example `join_game` without a loop would call `start_game` after any random event, such ass chat message or a new bot connecting. The loop is required to read all events until we receive the one breaking the loop.

Reply:
Okay I understand now, thanks for clarification

Reply:
Currently we are using synchronous socket.io implementation. That means that all calls are blocking --> CPU thread is stuck at given call waiting for the response.

There is a asynchronous implementation of socket.io client as well (it will be a part of later improvements of the client). The idea of async is that you create an async call and then `await` it. It might seems the same as sync (the code execution is stuck on the `await` line until the response from the server is received). However, the difference with async is that `await` signals to CPU that it is waiting for something. Based on that information CPU can temporarily assign this thread to a different process, until the interrupt triggering `await` is received and then puts the original awaiting process back on the thread (I don't think that it necessarily must be the same thread).

So yes, sync is kind of wasteful in a sense that it blocks the CPU thread (and CPU cannot assign it to any different process while the original one is waiting, because CPU does not know that the process is waiting). But during the waiting period, the thread does not compute anything, it is in idle state.

I hope it helps with understanding of the whole parallel computation a bit. Feel free to ask more questions.